Static Manifests Explicitly Target a K8s Version (for all supported versions)

afirth opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->

## What this PR does / why we need it:
Currently, the static manifests are only compatible with K8s 1.20 APIs. This change generates static manifests for each supported version by calling `helm template` with the recently re-added `--kube-version` flag

## Summary of changes
- besides the default K8s version (currently 1.20), static manifests for other supported versions are generated, fixing #7810
  - the various `semverCompare` statements are respected e.g. [this one](https://github.com/kubernetes/ingress-nginx/blob/30c0d2260d61fb942bb12ebf829e93dcb0b19988/charts/ingress-nginx/templates/default-backend-poddisruptionbudget.yaml#L2)
- add_namespace.py (which also deleted spec.replicas) is removed
  - the manual steps in [RELEASE](https://github.com/kubernetes/ingress-nginx/blob/helm-chart-4.0.12/RELEASE.md#e-edit-the-static-manifests) are not required and are removed because the generated yaml retains the "standard" K8s format
  - `helm --namespace` is used to set the namespace
  - `helm --kube-version` is used to set the target version
  - kustomize is used to delete spec.replicas and add the namespace resource
    - https://github.com/kubernetes/ingress-nginx/issues/4156 is not relevant as the generation is local
  - the inline heredocs in `generate-deploy-scripts.sh` and `add-namespace.py` move to yaml files whose directory layout matches that of the generated manifests

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [?] New feature (non-breaking change which adds functionality)
- [x] Documentation

## Which issue/s this PR fixes
fixes #7810

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
```bash
# check out a new branch from PR HEAD
gcb make-diffs

# checkout the manifests from main
git reset main -- ./deploy/static
git checkout main -- ./deploy/static

# deal with the one I moved
mkdir deploy/static/provider/aws/nlb-with-tls-termination
mv deploy/static/provider/aws/deploy-tls-termination.yaml deploy/static/provider/aws/nlb-with-tls-termination/deploy.yaml
cp deploy/static/provider/aws/kustomization.yaml deploy/static/provider/aws/nlb-with-tls-termination/

# run a hacky shell script which `kustomize builds` each directory using the existing kustomization.yaml
cat <<'EOF' > diffmaker.sh
#!/usr/bin/env bash
set -eux -o pipefail

#find . -type f -name deploy.yaml | xargs dirname
MANIFESTS=("./deploy/static/provider/do" "./deploy/static/provider/baremetal" "./deploy/static/provider/kind" "./deploy/static/provider/exoscale" "./deploy/static/provider/cloud" "./deploy/static/provider/scw" "./deploy/static/provider/aws" "./deploy/static/provider/aws/nlb-with-tls-termination/")

mkdir -p diffs
for dir in "${MANIFESTS[@]}"
do
  kustomize build $dir/ > diffs/$(basename $dir).yaml
done
EOF

chmod +x diffmaker.sh
./diffmaker.sh

ga diffmaker.sh
ga diffs
gcm 'original manifests'

# checkout the HEAD of the PR and generate those manifests
git reset 62e00e076 -- deploy/static
git checkout 62e00e076 -- deploy/static

ga diffs
gcm 'PR #8000 manifests'
gpsup
```

the diff is now visible between [7f3b490d1 and cfeda686a on my fork](https://github.com/afirth/ingress-nginx/commit/7f3b490d18aaa81bc5f71fad9a9f273afcd069e1) and shows that only the helm chart version annotation changed

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- n/a ~~I have added tests to cover my changes.~~
- [x] All new and existing tests passed.

## Detailed changes

dc420e604
- the manifests are regenerated at 4.0.12 (HEAD) using the existing script

85c487345
- the inline helm values in `hack/generate-deploy-scripts.sh` are replaced with files
- `hack/generate-deploy-scripts.sh` is DRYd with a loop over the values files
- the `deploy/static/provider/` directory is cleaned when `hack/generate-deploy-scripts.sh` is run, in case providers, variants, or K8s target versions are removed in the future

60c0f1e4a
- .tool-versions is added to ./hack for anyone using asdf (and generally to indicate to releasers or CI what versions to run)

1035cb9bc
- kustomize is used to mixin the namespace resource and python is removed
  - this results in a massive reordering, but no change besides the `spec.replicas` coming back for one commit
- the manual fix for `aws/deploy-tls-termination.yaml` specified in `RELEASE.md` is no longer required and is removed
- the `RELEASE_NAME` and `NAMESPACE` variables are removed from the shell script as they are only used once. Users wishing to customize the namespace can do so with kustomize by modifiying `namespace.yaml` and setting `namespace` in their overlay.
- the manifest `aws/deploy-tls-termination.yaml` is renamed to `aws/nlb-with-tls-termination/deploy.yaml`. This is the only provider with a variant manifest at this time. This change is necessary as multiple files are now used to generate the manifest (plus it smelled a bit)
- the `kustomization.yaml` present in each static manifest directory is generated from a template (previously created by hand)

33e5bff6e
- spec.replicas is removed where found (as previously done in `add_namespace.py`)

adcfd8be4
- static manifests are generated for all supported versions
- manifests targeting 1.20 (the [helm default](https://github.com/helm/helm/blob/v3.7.1/pkg/chartutil/capabilities.go#L31)) which was used to generate the existing manifests is retained at the root, so docs links will still work.

62e00e076
- the docs referencing the `aws/deploy-tls-termination.yaml` are updated with the new path
- `RELEASE.md` is updated to remove the manual steps and update dependencies
- trailing whitespace is cleaned in `RELEASE.md`
- quickstart docs updated with note about api version deprecations and that specific manifests are updated

## Notes to reviewers
Because this introduces reordering and formatting changes to the yaml files (because they are no longer loaded and dumped by python), I've left the intermediate commits and regenerated the manifests at each one, and recommend to review individually.

Additionally, I've loaded the manifests from `main` into kustomize and [generated diffs,](https://github.com/afirth/ingress-nginx/commit/7f3b490d18aaa81bc5f71fad9a9f273afcd069e1) so you can easily see what materially changed (just the chart version annotations)

## Further work

If desired I can make a github action to ensure that the static manifests are up to date for each PR

afirth commented:
/kind bug
ping @aledbf as the primary contributor of `generate-deploy-scripts.sh
/test all

tao12345666333 commented:
Thanks! I will finish the review today.

afirth commented:
> Thanks! I will finish the review today.

@tao12345666333 is there anything I can do to help this along?

aslafy-z commented:
Any chance this get rebased, reviewed and merged anytime soon @afirth? Thanks

aslafy-z commented on hack/generate-deploy-scripts.sh:
```diff
@@ -37,179 +37,23 @@ metadata:
-
-echo "${NAMESPACE_VAR}
-$(cat ${OUTPUT_FILE})" > ${OUTPUT_FILE}
+# clean
```
Shouldn't these changes go to another PR?

afirth commented:
> Any chance this get rebased, reviewed and merged anytime soon @afirth? Thanks

Rebased the docs (and fixed some more typos). Maybe reach out on slack to find reviewers? Or perhaps @caseydavenport can help? I'm happy to have a call with an OWNER and step through it if need be (@afirth on cncf slack)

tao12345666333 commented:
> @tao12345666333 is there anything I can do to help this along?

@afirth This is indeed a huge PR.I didn't spare so much time to finish it.

rikatz commented:
/assign

rikatz commented on RELEASE.md:
```diff
@@ -179,41 +179,18 @@ Promoting the images basically means that images, that were pushed to staging co
 
   - Prepare to use a script to update the edit the static manifests and set the "image", "digest", "version" etc. fields to the desired value.
 
+  - This script depends on kustomize and helm. The versions are pinned in `hack/.tool-versions` and you can use [asdf](https://github.com/asdf-vm/asdf#asdf) to install them
```
Would like not to use external installers in this case, but wont make this a show stopper. Just a follow up, it would be good if we just rely on internal scripts right now :)

afirth replied:
@rikatz previously the script depended on python, the ruamel.yaml python module, and an unspecified version of helm. There was no documentation about this that I saw.

Python is gone and replaced by a specific version of kustomize, and helm is also version pinned. `asdf` is purely a convenience that shims tools - like pyenv on steroids. If you install helm and kustomize in another way that's fine.

`asdf` does work well in CI, and it would be fairly easy I think to make a job that at least verified the generated files are up to date with any new commits. I'm happy to tackle that separately if you like. Anyway, I guess you are talking about asdf, which is optional. If you want me to update the docs to be more clear, or I misunderstood, please reply back and unresolve this.

rikatz commented on deploy/static/provider/aws/deploy-tls-termination.yaml:
```diff
@@ -38,11 +38,8 @@ metadata:
-      listen 2443;
-      return 308 https://$host$request_uri;
-    }
+  http-snippet: "server {\n  listen 2443;\n  return 308 https://$host$request_uri;\n\
```
This is still being generated in a broken state :/

afirth replied:
@rikatz I believe it's fixed in later commits. IIRC I had to modify the yaml in values, but the output now matches the manual output.
https://github.com/kubernetes/ingress-nginx/pull/8000/files#diff-00f078605badc375b2f2f8abbd7e676c1ce529d295f7fd7b214db190c16f8069R326

afirth replied:
(note that the file you're commenting on doesn't exist in the final commits)    .... i hope :laughing:

afirth commented:
/close
will bring in #8099 first, then trivial to modify the script to generate all versions as this one does.
cannot keep up with rebasing both
